pam/internal/adapter/nativemodel.go

```diff
@@ -0,0 +1,737 @@
+	qrCode, err := qrcode.New(m.uiLayout.GetContent(), qrcode.Medium)
+	if err != nil {
+		return sendEvent(pamError{
+			status: pam.ErrSystem,
+			msg:    fmt.Sprintf("Can't generate qrcode: %v", err),
+		})
+	}
+
+	if cmd := maybeSendPamError(m.sendInfo(m.uiLayout.GetLabel())); cmd != nil {
+		return cmd
+	}
+
+	qrcode := m.renderQrCode(qrCode)
+	if cmd := maybeSendPamError(m.sendInfo(qrcode)); cmd != nil {
+		return cmd
+	}
+
+	if code := m.uiLayout.GetCode(); code != "" {
+		firstLine := strings.SplitN(qrcode, "\n", 2)[0]
+		sizeDiff := len([]rune(firstLine)) - len(code)
+		var padding string
+		if sizeDiff > 0 {
+			padding = strings.Repeat(" ", sizeDiff/2)
+		}
+		if cmd := maybeSendPamError(m.sendInfo(padding + code + padding)); cmd != nil {
+			return cmd
+		}
+	}
+
+	// Ass some extra vertical space to improve readability
+	if cmd := maybeSendPamError(m.sendInfo(" ")); cmd != nil {
+		return cmd
+	}
+
+	choices := []choicePair{
+		{id: "wait", label: "Wait for the QR code scan result"},
+	}
+	if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+		choices = append(choices, choicePair{id: "button", label: buttonLabel})
+	}
+
+	id, err := m.promptForChoice("Qr Code authentication", choices, "Select action")
+	if errors.Is(err, errGoBack) {
+		return sendEvent(nativeGoBack{})
+	}
+	if errors.Is(err, errEmptyResponse) {
+		return sendAuthWaitCommand()
+	}
+	if err != nil {
+		return maybeSendPamError(err)
+	}
+
+	switch id {
+	case "button":
+		return sendEvent(reselectAuthMode{})
+	case "wait":
+		return sendAuthWaitCommand()
+	default:
+		return nil
+	}
+}
+
+func (m nativeModel) handleNewPassword() tea.Cmd {
+	if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+		choices := []choicePair{
+			{id: "continue", label: "Proceed with password update"},
+		}
+		if buttonLabel := m.uiLayout.GetButton(); buttonLabel != "" {
+			choices = append(choices, choicePair{id: "button", label: buttonLabel})
+		}
+
+		id, err := m.promptForChoice("Password Update", choices, "Select action")
+		if errors.Is(err, errGoBack) {
+			return sendEvent(nativeGoBack{})
+		}
+		if errors.Is(err, errEmptyResponse) {
+			return sendEvent(nativeChallengeRequested{})
+		}
+		if err != nil {
+			return maybeSendPamError(err)
+		}
+		if id == "button" {
+			return sendEvent(isAuthenticatedRequested{
+				item: &authd.IARequest_AuthenticationData_Skip{Skip: "true"},
+			})
+		}
+	}
+
+	return m.newPasswordChallenge(nil)
+}
+
+func (m nativeModel) newPasswordChallenge(previousChallenge *string) tea.Cmd {
+	if previousChallenge == nil {
+		if cmd := maybeSendPamError(m.sendInfo("Insert '%[1]s' to cancel the request and go back",
+			nativeCancelKey)); cmd != nil {
+			return cmd
+		}
+	} else {
+		if cmd := maybeSendPamError(m.sendInfo("Repeat the previously inserted password or insert '%[1]s' to cancel the request and go back",
+			nativeCancelKey)); cmd != nil {
+			return cmd
+		}
+	}
+
+	challenge, err := m.promptForChallenge(m.uiLayout.GetLabel())
```

Comment:
You just collected the challenge here. Why go back to the other model to validate it? Please validate it here before asking the user to confirm the challenge that might already be wrong anyway. I know you're worried about GDM, but GDM is another problem that we'll address in another pull request. Right now, doing the check here mimics the behavior of the TUI, and it's the best option, IMHO.

Reply:
The thing is that I need to:
 - Request the 1st challenge
 - Validate (asynchronously so that if some other event happens, we need to address it)
 - Get the 2nd challenge, and compare with the 1st

This implies:
 - Saving the old challenge, and that's done **already** by the authentication model that is our data model in this case
 - Save the 1st challenge (and we can just get it from the event).

All this can be done the same way inside this very same model, but I don't see why we should do the *very same thing* in three different UI models, when the `authentication.go` that is our logic backend can orchestrate all three without leaving this job to the UI, since it's not what it should do.

Reply:
The current abstraction is already quite flawed, but I don't think it's worth discussing this any further. So sure, let's keep it this way for now.